### ISSUE TITLE
请教关于Monodepth2一些问题

### ISSUE THREAD

**Issue by dexter2406** [kwea123/kwea123#5]
你好葵酱，我最近在用TF2复现monodepth2，代码层面基本搞定，但有些设定和训练细节我不太确定，故此询问一下！

**1. loss值的设置**
- 我观察到`total_loss` 的值非常低且变化范围非常小，这真的合理吗？从scratch训练也差不多就1.7，而官方给出的训练好的而模型loss差不多 0.05
- 我目前用预训练的resnet18想复现官方结果，但训练到epoch-10，total_loss也始终在1.1左右，且没有往下走的趋势，dispmap大概是如下的样子。感觉仍然是以 texture这种较低维度的特征为主，深度的层次不太明显（后面的小房子也被视作很近的物体了），跟官方的“完美结果”比起来差太远了。
- 请问你大概是训练了多少epoch 才复现官方结果呢？
![1](https://user-images.githubusercontent.com/46349448/116052433-e8af7a80-a679-11eb-8803-17f1cf9b10c6.png)
![image](https://user-images.githubusercontent.com/46349448/116052805-46dc5d80-a67a-11eb-9e2d-c73af3b61aa8.png)

**2. 训练设置**
- 官方代码中，`learning_rate=1e-4` 被设置为前 15 个 epoch的的值，且作者没有给出更多的解释，请问葵酱有没有一些调参的经验呢？比如前几个epoch 把 `lr`设置大一点，或者采用 *exponential* 的方式... （因为我的电脑有点菜，训练太慢了，验证idea 有点慢，所以想问问前辈的经验少走弯路）

**3. 关于 novel view synthesis**
- 我注意到在 inverse warping 之后，边缘像素会被按照原画边缘值补全`padding='border`，这样虽然比 `padding=zeros` 好一点，但仍然不是真实的像素值。是否有更好的方式呢？比如计算 loss 的时候加上mask，只计算中间 90% 的部分忽略边缘的模糊效应？
![2](https://user-images.githubusercontent.com/46349448/116054262-c7e82480-a67b-11eb-9864-8c286691f770.png)

谢谢 :)

**Comment by kwea123** [comment]
你好，
1. 我并没有尝试复现官方结果，我只有使用他预设的参数训练过apolloscape的一些图像，当时并未发现有loss不下降的状况，所以我没有深入探讨。当时的loss下降状况我也忘了。。所以关于这部分我想我没办法给你太多建议。
2. learning rate我觉得应该不是影响最大的，我觉得应该是loss还是warping的地方出了问题吧？一般来说，learning rate都是exponential或是cosine之类随着epoch的进行调小会让精度有所上升，这点是没错的，基本上可以直接这样设定。
3. 这个问题我有问过。https://github.com/nianticlabs/monodepth2/issues/20 最好的方式是要用mask然后padding=zeros才是对的。

**Comment by dexter2406** [comment]
@kwea123 谢谢你的意见！关于第3点我也是这么想的，我已经改好了代码，到时候瞅瞅结果如何。
我发现自己的问题出在哪里了——官方代码将输入值做了个 `x = (x - 0.45) / 0.225` 的运算，而我没有，所以结果总是怪怪的，我改过来之后发现其实已经训练的不错了！

所以我还想问一下，这个 `x = (x - 0.45) / 0.225` 到底有什么作用呢，我看到很多模型里都对*ImageNet*图片做了这样的预处理，但不知道作用是什么...
![test_image20210426-171529](https://user-images.githubusercontent.com/46349448/116107201-1cf35d00-a6b3-11eb-8ec4-641389e6227f.jpg)

**Comment by dexter2406** [comment]
我查了下资料好像明白了，这个操作是为了将 input 正则化到 (-1, 1)，还是 (0.5, 0.5)来着
既然这样，我觉得有点多此一举了—— 0.45 和 0.255 是基于对kitti 数据集已知的情况，而如果用于未知图像，这个值就不一样了，那还不如不加，毕竟Encoder里的 BN层是可以通过 `running_mean`和 `running_var` 来正则化图片的。不知道我说的对不对，期待您的意见！

**Comment by kwea123** [comment]
對於input的正則化通常都是手動（或計算data的mean, std)來決定的。這部份比較算是一個common practice。你說的用bn當然也可以（是做相同的事），但只是沒人這麼做而已。一般來說用imagenet pretrain的network會使用imagenet的mean跟var做normalize，這邊它用的0.45跟0.255其實是imagenet的mean跟var的近似值。
mean跟var是跟數據集有關沒錯，但實際他的數字並沒有很重要，重要的是必須做這個動作，因為network對(-1, 1)的input會比較好訓練。

**Comment by Shaw-Way** [comment]
> @kwea123 谢谢你的意见！关于第3点我也是这么想的，我已经改好了代码，到时候瞅瞅结果如何。 我发现自己的问题出在哪里了——官方代码将输入值做了个 `x = (x - 0.45) / 0.225` 的运算，而我没有，所以结果总是怪怪的，我改过来之后发现其实已经训练的不错了！
> 
> 所以我还想问一下，这个 `x = (x - 0.45) / 0.225` 到底有什么作用呢，我看到很多模型里都对_ImageNet_图片做了这样的预处理，但不知道作用是什么... ![test_image20210426-171529](https://user-images.githubusercontent.com/46349448/116107201-1cf35d00-a6b3-11eb-8ec4-641389e6227f.jpg)

hello，请问您前述的mask是怎么做的
